stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+    check_range(solver_->function_name_, "interpolation_polynomial", 2,
+                interpolation_polynomial_);
+    // 1=Adams=CV_ADAMS, 2=BDF=CV_BDF
+    check_range(solver_->function_name_, "solver_forward", 2, solver_forward_);
+    check_range(solver_->function_name_, "solver_backward", 2,
+                solver_backward_);
+
+    check_flag_sundials(
+        CVodeInit(solver_->cvodes_mem_, &cvodes_integrator_adjoint_vari::cv_rhs,
+                  value_of(t0_), solver_->nv_state_forward_),
+        "CVodeInit");
+
+    // Assign pointer to this as user data
+    check_flag_sundials(
+        CVodeSetUserData(solver_->cvodes_mem_, reinterpret_cast<void*>(this)),
+        "CVodeSetUserData");
+
+    cvodes_set_options(solver_->cvodes_mem_, relative_tolerance_forward_,
+                       absolute_tolerance_forward_(0), max_num_steps_);
+
+    check_flag_sundials(
+        CVodeSVtolerances(solver_->cvodes_mem_, relative_tolerance_forward_,
+                          solver_->nv_absolute_tolerance_forward_),
+        "CVodeSVtolerances");
+
+    check_flag_sundials(
+        CVodeSetLinearSolver(solver_->cvodes_mem_, solver_->LS_forward_,
+                             solver_->A_forward_),
+        "CVodeSetLinearSolver");
+
+    check_flag_sundials(
+        CVodeSetJacFn(solver_->cvodes_mem_,
+                      &cvodes_integrator_adjoint_vari::cv_jacobian_states),
+        "CVodeSetJacFn");
+
+    // initialize backward sensitivity system of CVODES as needed
+    if (is_var_return) {
+      check_flag_sundials(
+          CVodeAdjInit(solver_->cvodes_mem_, num_steps_between_checkpoints_,
+                       interpolation_polynomial_),
+          "CVodeAdjInit");
+    }
+  }
+
+  /**
+   * Solve the ODE initial value problem y' = f(t, y), y(t0) = y0 at a set of
+   * times, { t1, t2, t3, ... } using the stiff backward differentiation formula
+   * (BDF) solver in CVODES.
+   *
+   * @return std::vector of Eigen::Matrix of the states of the ODE, one for each
+   *   solution time (excluding the initial state)
+   */
+  std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> operator()() {
+    const double t0_dbl = value_of(t0_);
+    const auto ts_dbl = value_of(ts_);
+    std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> y_return(
+        ts_.size());
+
+    double t_init = t0_dbl;
+    for (size_t n = 0; n < ts_dbl.size(); ++n) {
+      double t_final = ts_dbl[n];
+      if (t_final != t_init) {
+        if (is_var_return) {
+          int ncheck;
+
+          int error_code
+              = CVodeF(solver_->cvodes_mem_, t_final,
+                       solver_->nv_state_forward_, &t_init, CV_NORMAL, &ncheck);
+
+          if (error_code == CV_TOO_MUCH_WORK) {
+            throw_domain_error(solver_->function_name_, "", t_final,
+                               "Failed to integrate to next output time (",
+                               ") in less than max_num_steps steps");
```

Comment:
Since this is done in a loop we might want to do the [`STAN_COLD_PATH`](https://github.com/stan-dev/math/blob/80f8578e1e531d0467d9233ac158a4419fec1ee4/stan/math/prim/err/check_range.hpp#L29) trick to let the compiler know this branch almost never happens

Reply:
Maybe, but in this case it's not so urgent, I think. The thing is that integrating from one to the next time-step takes forever relative to these if statements.